File: src/Curve/E521.java
package Curve;
import java.math.BigInteger;

/**
 * Curve.E521 Elliptic Curve (Edward's Curve) of equation: (x^2) + (y^2) = 1 + d(x^2)(y^2)
 * where d = -376014
 * Contains methods to add and multiply points on curve using scalar values.
 *
 * @version Spring 2021
 * @author Dustin Ray
 */
public class E521 {

    /** Mersenne prime defining a finite field F(p) */
    public static final BigInteger P = (new BigInteger("2").pow(521)).subtract(BigInteger.ONE);
    /** d = -376014 */
    public static final BigInteger D = new BigInteger("-376014");
    /** X-coordinate of point. */
    private BigInteger myX;
    /** Y-coordinate of point. */
    private BigInteger myY;
    /** number of points on Curve.E521 Curve -> n := 4 * (R) .*/
    public static final BigInteger R = (new BigInteger("2").pow(519))
            .subtract(new BigInteger(
                    "337554763258501705789107630418782636071904961214051226618635150085779108655765"));

    /**
     * Constructor for point with only X provided. Solves for Y.
     * @param theX int value which is desired X coordinate of point. Y is generated
     * from X using sqrt method.
     */
    public E521(final BigInteger theX, boolean theLSB) {
        myX = theX;
        //solve for y
        BigInteger num = (BigInteger.ONE.subtract(theX.pow(2))).mod(P);
        BigInteger denom = BigInteger.ONE.add(new BigInteger("376014").multiply(theX.pow(2))).mod(P);
        denom = denom.modInverse(P);
        BigInteger radicand = num.multiply(denom);
        myY = sqrt(radicand, P, theLSB);
    }

    /**
     * Constructor for point with arbitrary x and y.
     * @param theX x coordinate of point
     * @param theY y coordinate of point
     */
    public E521(final BigInteger theX, final BigInteger theY) {
        myX = theX;
        myY = theY;
    }

    /** Constructor for neutral element, defined to be point (0, 1) */
    public E521() {
        myX = BigInteger.ZERO;
        myY = BigInteger.ONE;
    }

    /**
     * Compares two points for equality by X and Y coordinates.
     * Utilizes .equals method of BigInteger class.
     *
     * @param theOther Curve.E521 point to be compared against this object.
     * @return boolean, true for equal, else otherwise.
     */
    public boolean equals(final E521 theOther) {
        return this.getX().equals(theOther.getX())
                && this.getY().equals(theOther.getY());
    }

    /**
     * Gets the opposite value of a point, defined as the following:
     * if P = (X, Y), opposite of P = (-X, Y).
     *
     * @return Curve.E521 point which is opposite of a given point as defined above.
     */
    public E521 getOpposite() {
        return new E521(myX.negate(), this.getY());
    }

    /**
     * Adds two Curve.E521 points and returns another (X, Y) point which is on the Curve.E521 curve defined
     * using parameters set by constructor. Add operation is defined as:
     * (x1, y1) + (x2, y2) = ((x1y2 + y1x2) / (1 + (d)x1x2y1y2)), ((y1y2 - x1x2) / (1 - (d)x1x2y1y2))
     * where "/" is defined to be multiplication by modular inverse.
     *
     * @param theOther e521 point to be added to current point.
     * @return Curve.E521 point which is the result of the addition operation defined here. Will
     * fail with exception if point is not on curve defined in class fields.
     */
    public E521 add(final E521 theOther) {

        BigInteger x1 = this.getX();
        BigInteger x2 = theOther.getX();

        BigInteger y1 = this.getY();
        BigInteger y2 = theOther.getY();

        BigInteger xNum = ((x1.multiply(y2)).add(y1.multiply(x2))).mod(P);
        BigInteger xDenom = (BigInteger.ONE.add(D.multiply(x1).multiply(x2).multiply(y1).multiply(y2))).mod(P);

        xDenom = xDenom.modInverse(P);
        BigInteger newX = xNum.multiply(xDenom).mod(P);

        BigInteger yNum = (y1.multiply(y2).subtract(x1.multiply(x2))).mod(P);
        BigInteger yDenom = (BigInteger.ONE.subtract(D.multiply(x1).multiply(x2).multiply(y1).multiply(y2))).mod(P);

        yDenom = yDenom.modInverse(P);
        BigInteger newY = yNum.multiply(yDenom).mod(P);

        return new E521(newX, newY);
    }

    /**
     * Naive multiplication. Computes sP as P[1] + P[2] ... P[i] 
     * @param theS is number of times to add P to itself. 
     * @param P is the point to multiply. 
     */
    public E521 multiply(final BigInteger theS, final E521 P) {
        E521 result = new E521(BigInteger.ZERO, BigInteger.ONE);
        for(int i = 0; i < theS.intValue(); i ++) {
            result = result.add(P);
        }
        return result;
    }

    /**
     * Double and add multiplication for Curve.E521 point, index decreasing.
     * @param theS Scalar value to multiply by
     * @return Curve.E521 point which is product of multiplication
     * with S.
     */
    public E521 doubleAndAdd(final BigInteger theS) {

        E521 result = new E521(BigInteger.ZERO, BigInteger.ONE);
        int idx = theS.bitLength();
        while (idx >= 0) {
            result = result.add(result);
            if (theS.testBit(idx--)) {
                result = result.add(this);
            }
        }
        return result;
    }

    /**
     * EC Multiplication algorithm using the Montgomery Ladder approach to mitigate
     * timing side channel attacks. Mostly constructed around
     * https://eprint.iacr.org/2014/140.pdf pg 4
     * 2R here is defined as a call to the addition method to act as a doubling algorithm. Can probably
     * be replaced with individual doubling algorithm.
     *
     * @param theS scalar value to multiply by. S is a private key and should be kept secret.
     * @return Curve.E521 point which is result of multiplication.
     */
    public E521 multiplyMontgomery(final BigInteger theS, final E521 P) {

        E521 r0 = new E521(BigInteger.ZERO, BigInteger.ONE);
        E521 r1 = P;
        int idx = theS.bitLength();
        while (idx >= 0) {
            if (theS.testBit(idx--)) {
                r0 = r0.add(r1);
                r1 = r1.add(r1);
            } else {
                r1 = r0.add(r1);
                r0 = r0.add(r0);
            }
        }
        return r0; // r0 = P * s
    }


    /**
     * Compute a square root of v mod p with a specified
     * least significant bit, if such a root exists.
     *
     * @param v   the radicand.
     * @param p   the modulus (must satisfy p mod 4 = 3).
     * @param lsb desired least significant bit (true: 1, false: 0).
     * @return a square root r of v mod p with r mod 2 = 1 iff lsb = true
     * if such a root exists, otherwise null.
     */
    public static BigInteger sqrt(BigInteger v, BigInteger p, boolean lsb) {
        assert (p.testBit(0) && p.testBit(1)); // p = 3 (mod 4)
        if (v.signum() == 0) {
            return BigInteger.ZERO;
        }
        BigInteger r = v.modPow(p.shiftRight(2).add(BigInteger.ONE), p);
        if (r.testBit(0) != lsb) {
            r = p.subtract(r); // correct the lsb }
            return (r.multiply(r).subtract(v).mod(p).signum() == 0) ? r : null;
        }
        return r;
    }


    /**
     * Gets string representation of object as pair of X and Y coordinates.
     * @return String which is labelled pair of X and Y coordinates.
     */
    public String toString(final String name) {
        return name + "\n" + "X coordinate: " + getX().toString() + "\n"
                + "Y coordinate: " + getY().toString() + "\n";
    }

    /**
     * Gets current X value.
     * @return BigInteger which is current X coordinate of point.
     */
    public BigInteger getX() {
        return myX;
    }

    /** Set X coordinate. */
    public void setX(final BigInteger theX) {myX = theX;}
    /** Set Y coordinate. */
    public void setY(final BigInteger theY) {myY = theY;}

    /**
     * Gets current Y value.
     * @return BigInteger which is current Y coordinate of point.
     */
    public BigInteger getY() { return myY; }

    /** Returns r value for curve.
     * @return r value for curve. */
    public BigInteger getR() { return R; }




}

File: e521/src/curve/e521.rs
pub mod e521 {
    use std::{str::FromStr, ops::{Sub, Mul}};
    use num_bigint::{BigInt, Sign};
    use num::Integer;
    use num::One;
    use num::Zero;
    /*
    The elliptic curve that will be implemented is known as the 𝐸521 curve (a so-
    called Edwards curve), defined by the following parameters:
    • 𝑝 ≔ 2^521−1, a Mersenne prime defining the finite field 𝔽𝑝 .
    • curve equation: 𝑥^2 + 𝑦^2 = 1 + 𝑑𝑥^2𝑦^2 with 𝑑 = −376014. 
    */
    #[derive(Debug)]
    pub struct E521 {
    pub x: BigInt,
    pub y: BigInt,
    pub p: BigInt,
    pub d: BigInt,
    pub r: BigInt,
    pub n: BigInt,
    }
    /// Initializes r value for curve. 
    fn set_r() -> BigInt {
    let r = BigInt::from(2);
    let r = r.pow(519);
    let s = BigInt::from_str("337554763258501705789107630418782636071904961214051226618635150085779108655765").unwrap();
    let r = r.sub(s);
    return r;
    }

    /// Initializes curve order, 𝑝 ≔ 2^521−1, a Mersenne prime defining the finite field 𝔽𝑝.
    fn set_p() -> BigInt {
    let p = BigInt::from(2);
    let p = p.pow(521);
    let p = p - 1;
    return p;
    }

    /// Initializes number of points on the curve. 
    fn set_n() -> BigInt {
    let n = set_r();
    let n = n * &BigInt::from(4);
    return n;
    }

    /// Sets the curve d parameter. 
    fn set_d() -> BigInt { return Some(BigInt::from(-376014)).unwrap();}





    /// Generates the neutral point (0, 1)
    pub fn get_e521_id_point() -> E521 {
    let point = E521{
        x: BigInt::from(0),
        y: BigInt::from(1),
        p: set_p(),
        d: set_d(),
        r: set_r(), 
        n: set_n()
    };
    point
    }

    /// Gets point for arbitrary (x, y) TODO verify point is on curve
    pub fn get_e521_point(x: BigInt, y: BigInt) -> E521 {
    let point = E521{
        x,
        y,
        p: set_p(),
        d: set_d(),
        r: set_r(), 
        n: set_n()
    };
    point
    }

    /// Gets point for arbitrary (x, y) TODO verify point is on curve
    pub fn get_e521_gen_point(msb: bool) -> E521 {
    let x = BigInt::from(4);
    let new_x = x.clone();
    let point = E521{
        x,
        y: solve_for_y(&new_x, set_p(), msb),
        p: set_p(),
        d: set_d(),
        r: set_r(), 
        n: set_n()
    };
    point
    }

    /// If a point is defined as (x, y) then its negation is (-x, y)
    pub fn negate_point(p: &E521) -> E521 {
    let x = p.x.clone();
    let y = p.y.clone();
    let x = x * -1;
    let point = get_e521_point(x, y);
    point
    }


    // point operations


    ///Adds two E521 points and returns another E521 curve point. If a point is defined as
    /// E521 = (x, y), then E521 addition is defined as:

    ///(x1, y1) + (x2, y2) = ((x1y2 + y1x2) / (1 + (d)x1x2y1y2)), ((y1y2 - x1x2) / (1 - (d)x1x2y1y2))

    ///where "/" is defined to be multiplication by modular inverse.
    pub fn add_points(p1  : &E521, p2: &E521) -> E521 {

    let p = p1.p.clone();
    let d = p1.d.clone();

    let x1 = p1.x.clone();
    let y1 = p1.y.clone();

    // (x1y2 + y1x2)
    let x1y2 = (x1 * &p2.y).mod_floor(&p);
    let y1x2 = (y1 * &p2.x).mod_floor(&p);
    let x1y2y1x2_sum = (x1y2 + y1x2).mod_floor(&p);

    let x1 = p1.x.clone();

    // 1 / (1 + (d)x1x2y1y2))
    let one_plus_dx1x2y1y2 = (BigInt::from(1) + (d * x1 * &p2.x) *(&p1.y)* &p2.y).mod_floor(&p);
    let one_plus_dx1x2y1y2inv = mod_inv(&one_plus_dx1x2y1y2, &p);

    let x1 = p1.x.clone();
    let y1 = p1.y.clone();

    // (y1y2 - x1x2)
    let y1y2x1x2_difference = ((y1 * &p2.y) - (x1 * &p2.x)).mod_floor(&p);

    let x1 = p1.x.clone();
    let d = p1.d.clone();

    // 1 / (1 - (d)x1x2y1y2))
    let one_minus_dx1x2y1y2 = (BigInt::from(1) - (d * x1 * &p2.x * &p1.y * &p2.y)).mod_floor(&p);
    let one_minus_dx1x2y1y2inv = mod_inv(&one_minus_dx1x2y1y2, &p);

    let new_x = (x1y2y1x2_sum * one_plus_dx1x2y1y2inv).mod_floor(&p);
    let new_y = (y1y2x1x2_difference * one_minus_dx1x2y1y2inv).mod_floor(&p);;
    get_e521_point(new_x, new_y)
    }

    /// constant time multiplication NOTE not memory safe afaik.
    /// multiplication is defined to be P + P + P .... s times
    pub fn sec_mul(s: &BigInt, point: &E521) -> E521 {
    let mut r1 = get_e521_gen_point(false);
    let mut r0 = get_e521_id_point();
    let len = s.bits();
    for i in (0..len).rev()  {
        if s.bit(i) {
            r0 = add_points(&r0, &r1);
            r1 = add_points(&r1, &r1);
        } else { 
            r1 = add_points(&r0, &r1);
            r0 = add_points(&r0, &r1);
        }
        }
    r0
    }



    /// Solves for y in curve equation 𝑥^2 + 𝑦^2 = 1 + 𝑑𝑥^2𝑦^2
    fn solve_for_y(x: &BigInt, p: BigInt, msb: bool) -> BigInt {
    let num = BigInt::from(1) - x.pow(2);
    let num = mod_formula(&num, &p);
    let denom = BigInt::from(376_014) * x.pow(2) + BigInt::from(1);
    let denom = mod_formula(&denom, &p);
    let denom = mod_inv(&denom, &p);
    let radicand = num * denom;
    let y = sqrt(&radicand, p, msb);
    y
    }

    /// Compute a square root of v mod p with a specified
    /// least significant bit, if such a root exists.
    fn sqrt(v: &BigInt, p: BigInt, lsb: bool) -> BigInt {
    if v.sign() ==  Sign::NoSign{ return BigInt::from(0); }
    let r = v.modpow(&((p.clone() >> 2) + 1), &p);
    if !r.bit(0).eq(&lsb) {
        let new_r = &p - r; // correct the lsb
        let borrowed_r = new_r.clone();
        let return_r = new_r.clone();
        let bi = mod_formula(&new_r.mul(borrowed_r).sub(v), &p);
        if bi.sign() == Sign::NoSign {
            return return_r;
        } else { return BigInt::from(0); }
    } r
    }

    /// Performs BigInt modular arithematic.
    fn mod_formula(a: &BigInt, b: &BigInt) -> BigInt { ((a % b) + b) % b }

    /// Performs modular inverse via euclidian algorithm. 
    fn mod_inv(n: &BigInt, p: &BigInt) -> BigInt {
    if p.is_one() { return BigInt::one() }
    let (mut a, mut m, mut x, mut inv) = (n.clone(), p.clone(), BigInt::zero(), BigInt::one());
    while a < BigInt::zero() { a += p }
    while a > BigInt::one() {
        let (div, rem) = a.div_rem(&m);
        inv -= div * &x;
        a = rem;
        std::mem::swap(&mut a, &mut m);
        std::mem::swap(&mut x, &mut inv);
    }
    if inv < BigInt::zero() { inv += p }
    inv
    }


}